runtime-sdk/modules/evm/src/backend.rs

```diff
@@ -1,330 +1,526 @@
+pub struct OasisBackend<'ctx, C: TxContext, Cfg: Config> {
     vicinity: Vicinity,
     ctx: RefCell<&'ctx mut C>,
+    pending_state: RefCell<context::State>,
     _cfg: PhantomData<Cfg>,
 }
 
-impl<'ctx, C: Context, Cfg: Config> Backend<'ctx, C, Cfg> {
+impl<'ctx, C: TxContext, Cfg: Config> OasisBackend<'ctx, C, Cfg> {
     pub fn new(ctx: &'ctx mut C, vicinity: Vicinity) -> Self {
         Self {
             vicinity,
             ctx: RefCell::new(ctx),
+            pending_state: RefCell::new(context::State::default()),
             _cfg: PhantomData,
         }
     }
 }
 
-impl<'ctx, C: Context, Cfg: Config> EVMBackend for Backend<'ctx, C, Cfg> {
-    fn gas_price(&self) -> primitive_types::U256 {
-        self.vicinity.gas_price.into()
+/// An extension trait implemented for any [`Backend`].
+pub(crate) trait EVMBackendExt {
+    /// Returns at most `num_bytes` bytes of cryptographically secure random bytes.
+    /// The optional personalization string may be included to increase domain separation.
+    fn random_bytes(&self, num_bytes: u64, pers: &[u8]) -> Vec<u8>;
+
+    /// Perform a subcall.
+    fn subcall<V: subcall::Validator + 'static>(
+        &self,
+        info: subcall::SubcallInfo,
+        validator: V,
+    ) -> Result<subcall::SubcallResult, core::Error>;
+}
+
+impl<T: EVMBackendExt> EVMBackendExt for &T {
+    fn random_bytes(&self, num_bytes: u64, pers: &[u8]) -> Vec<u8> {
+        (*self).random_bytes(num_bytes, pers)
+    }
+
+    fn subcall<V: subcall::Validator + 'static>(
+        &self,
+        info: subcall::SubcallInfo,
+        validator: V,
+    ) -> Result<subcall::SubcallResult, core::Error> {
+        (*self).subcall(info, validator)
+    }
+}
+
+impl<'ctx, C: TxContext, Cfg: Config> EVMBackendExt for OasisBackend<'ctx, C, Cfg> {
+    fn random_bytes(&self, num_bytes: u64, pers: &[u8]) -> Vec<u8> {
+        // Refuse to generate more than 1 KiB in one go.
+        // EVM memory gas is checked only before and after calls, so we won't
+        // see the quadratic memory cost until after this call uses its time.
+        let num_bytes = num_bytes.min(RNG_MAX_BYTES) as usize;
+        let mut ctx = self.ctx.borrow_mut();
+        let mut rng = ctx.rng(pers).expect("unable to access RNG");
+        let mut rand_bytes = vec![0u8; num_bytes];
+        rand_core::RngCore::try_fill_bytes(&mut rng, &mut rand_bytes).expect("RNG is inoperable");
+        rand_bytes
+    }
+
+    fn subcall<V: subcall::Validator + 'static>(
+        &self,
+        info: subcall::SubcallInfo,
+        validator: V,
+    ) -> Result<subcall::SubcallResult, core::Error> {
+        let mut ctx = self.ctx.borrow_mut();
+
+        // Execute the subcall.
+        let mut result = subcall::call(&mut ctx, info, validator)?;
+        // Store state after subcall execution for substate handling.
+        self.pending_state
+            .borrow_mut()
+            .merge_from(mem::take(&mut result.state));
```

Comment:
Which docs?

Reply:
```
    /// State after applying the subcall context.
    pub state: State,
```

Reply:
Ideally one would have access to the substate from the precompile. Unfortunately this is not the case so such an ugly hack is needed where stuff meant for the substate is stored in the backend and then pulled into substate.

Alternatively we could just `clone` (seems useless though as the caller cannot do anything useful with this state), return multiple components or introduce a separate type that doesn't have the `state` field.

Reply:
> ugly hack is needed

I saw that, hack is fine.

> Alternatively we could just clone.

That would be a solution. Maybe we just add some kind of a comment, so that if we add another caller in the future he will be aware that the state could be cleared.

Reply:
A comment has been added.